File: Java/0022_generate_parentheses/0022_generate_parentheses.java
class Solution {
    public List<String> generateParentheses(int n) {
        List<String> result = new ArrayList<String>();
        addParentheses(result, "", n, 0);
        return result;
    }
    
    public void addParentheses(List<String> result, String str, int notAdd, int notMatch) {
        if (notMatch == 0 && notAdd == 0) {
            result.add(str);
            return;
        }
        if (notMatch > 0)
            addParentheses(result, str + ')', notAdd, notMatch - 1);
        if (notAdd > 0)
            addParentheses(result, str + '(', notAdd - 1, notMatch + 1);
    }  
}

File: Python/0022_generate_parentheses/0022_generate_parentheses.py
class Solution:
    def generateParentheses(self, n):
        """
        :type n: int
        :rtype: List[str]
        """
        result = []
        self.addParentheses(result, "", n, 0)
        return result
        
    def addParentheses(self, result, string, notAdd, notMatch):
        if notMatch == 0 and notAdd == 0:
            result.append(string)
            return
        if notMatch > 0:
            self.addParentheses(result, string + ')', notAdd, notMatch - 1)
        if notAdd > 0:
            self.addParentheses(result, string + '(', notAdd - 1, notMatch + 1)

